distributed/tests/test_worker_state_machine.py

```diff
@@ -1351,3 +1352,180 @@ def test_transfer_incoming_metrics(ws):
     assert ws.transfer_incoming_bytes == 0
     assert ws.transfer_incoming_count == 0
     assert ws.transfer_incoming_count_total == 4
+
+
+def test_throttling_does_not_affect_first_transfer(ws):
+    ws.transfer_incoming_count_limit = 100
+    ws.transfer_incoming_bytes_limit = 100
+    ws.transfer_incoming_bytes_throttle_threshold = 1
+    ws2 = "127.0.0.1:2"
+    ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            "c",
+            who_has={"a": [ws2]},
+            nbytes={"a": 200},
+            stimulus_id="s1",
+        )
+    )
+    assert ws.tasks["a"].state == "flight"
+
+
+def test_throttle_incoming_transfers_on_count_limit(ws):
+    ws.transfer_incoming_count_limit = 1
+    ws.transfer_incoming_bytes_limit = 100_000
+    ws.transfer_incoming_bytes_throttle_threshold = 1
+    ws2 = "127.0.0.1:2"
+    ws3 = "127.0.0.1:3"
+    ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            "c",
+            who_has={"a": [ws2], "b": [ws3]},
+            nbytes={"a": 100, "b": 100},
+            stimulus_id="s1",
+        )
+    )
+    assert ws.tasks["a"].state == "flight"
+    assert ws.tasks["b"].state == "fetch"
+    assert ws.transfer_incoming_bytes == 100
+
+    ws.handle_stimulus(
+        GatherDepSuccessEvent(
+            worker=ws2, data={"a": 123}, total_nbytes=100, stimulus_id="s2"
+        )
+    )
+    assert ws.tasks["b"].state == "flight"
```

Comment:
Similar concern about determinism
```suggestion
    tasks_by_state = defaultdict(list)
    for ts in ws.tasks.values():
        tasks_by_state[ts.state].append(ts)
    assert len(tasks_by_state["flight"]) == 1
    assert len(tasks_by_state["fetch"]) == 1
    assert ws.transfer_incoming_bytes == 100

    ws.handle_stimulus(
        GatherDepSuccessEvent(
            worker=ws2,
            data={tasks_by_state["flight"][0]: 123},
            total_nbytes=100,
            stimulus_id="s2",
        )
    )
    assert ws.tasks[tasks_by_state["fetch"][0]].state == "flight"
```